lib/user.rb

```diff
@@ -0,0 +1,29 @@
+require 'bcrypt'
+class User
+
+  include DataMapper::Resource
+
+  attr_reader :password
+  attr_accessor :password_confirmation
+
+  validates_confirmation_of :password
+
+  property :id, Serial
+  property :email, String,
+           unique: true,
+           message: 'This email address has be taken'
+  property :password_digest, Text
+  property :name, String, unique: true, message: 'This name has be taken'
+  property :username, String
+
+  def password=(password)
+    @password = password
+    self.password_digest = BCrypt::Password.create(password)
+  end
+
+  def self.authenticate(email, password)
+    user = first(email: email)
+    user if user && BCrypt::Password.new(user.password_digest) == password
+  end
+
+end
```

Comment:
let's take the 
user:  he can have many peeps and many replies
peep: can belong only to one user and can have many replies(which can be from different users)
reply: belongs to only one user and only to one reply at the same time
Data mapper provides a way to this(instead of 'though: Resource can be through another model(like 'peep')')